rts/Lua/LuaHandleSynced.cpp

```diff
@@ -2120,16 +2120,187 @@ int CSyncedLuaHandle::GetWatchWeaponDef(lua_State* L) {
+ */
+
 GetWatchDef(Feature)
+
+
+/*** Query general watch status for a weaponDefID.
+ *
+ * @function Spring.GetWatchWeapon
+ *
+ * Same as calling:
+ * ```lua
+ * Spring.GetWatchExplosion(weaponDefID) or Spring.GetWatchProjectile(weaponDefID) or Spring.GetWatchAllowTarget(weaponDefID)
+ * ```
+ *
+ * @param weaponDefID integer
+ * @return boolean watched True if watch is enabled for any weaponDefID callins.
+ *
+ * @see Spring.SetWatchWeapon
+ */
+
+/*** Query Explosion watch status for a weaponDefID.
+ *
+ * @function Spring.GetWatchExplosion
+ *
+ * @param weaponDefID integer
+ * @return boolean watched Watch status.
+ *
+ * @see Spring.SetWatchExplosion
+ */
+
 GetWatchDef(Explosion)
+
+
+/*** Query Projectile watch status for a weaponDefID.
+ *
+ * @function Spring.GetWatchProjectile
+ *
+ * @param weaponDefID integer
+ * @return boolean watched Watch status.
+ *
+ * @see Spring.SetWatchProjectile
+ */
+
 GetWatchDef(Projectile)
+
+
+/*** Query AllowTarget watch status for a weaponDefID.
+ *
+ * @function Spring.GetWatchAllowTarget
+ *
+ * @param weaponDefID integer
+ * @return boolean watched Watch status.
+ *
+ * @see Spring.SetWatchAllowTarget
+ */
+
 GetWatchDef(AllowTarget)
 
+
+/*** Register or deregister unitDefID for expensive callins.
+ *
+ * @function Spring.SetWatchUnit
+ *
+ * @param unitDefID integer
+ * @param watch boolean Whether to register or deregister.
+ * @return nil
+ *
+ * @see Callins:UnitFeatureCollision
+ * @see Callins:UnitUnitCollision
+ * @see Callins:UnitMoveFailed
+ */
+
 SetWatchDef(Unit)
+
+
+/*** Register or deregister featureDefID for expensive callins.
+ *
+ * @function Spring.SetWatchFeature
+ *
+ * @param featureDefID integer
+ * @param watch boolean Whether to register or deregister.
+ * @return nil
+ *
+ * @see Callins:UnitFeatureCollision
+ */
+
 SetWatchDef(Feature)
+
+
+/*** Register or deregister weaponDefID for all expensive callins.
+ *
+ * @function Spring.SetWatchWeapon
+ *
+ * Equivalent to calling:
+ *
+ * ```lua
+ * Spring.SetWatchExplosion(weaponDefID)
+ * Spring.SetWatchProjectile(weaponDefID)
+ * Spring.SetWatchAllowTarget(weaponDefID)
+ * ```
+ *
+ * Generally it's better to use those methods to avoid registering uneeded callins.
+ *
+ * @param weaponDefID integer
+ * @param watch boolean Whether to register or deregister.
+ * @return nil
+ */
+
+/*** Register or deregister weaponDefID for expensive Explosion callins.
+ *
+ * @function Spring.SetWatchExplosion
+ *
+ * @param weaponDefID integer
+ * @param watch boolean Whether to register or deregister.
+ * @return nil
+ *
+ * @see SyncedCallins:Explosion
+ */
+
 SetWatchDef(Explosion)
+
+
+/*** Register or deregister weaponDefID for expensive Projectile callins.
+ *
+ * @function Spring.SetWatchProjectile
+ *
+ * @param weaponDefID integer weaponDefID for weapons or -1 to watch for debris.
+ * @param watch boolean Whether to register or deregister.
+ * @return nil
+ *
+ * @see SyncedCallins:ProjectileCreated
```

Comment:
```suggestion
 * @see Callins:ProjectileCreated
```

https://github.com/beyond-all-reason/RecoilEngine/blob/0683321af751bf99a7fe3fc28f5e1ca76b80c591/rts/Lua/LuaHandle.cpp#L2076

Reply:
you sure about this? I seen SyncedCallins in other places and these are indeed synced, otherwise if every callin has to go in Callins its fine with me too.

Reply:
`ProjectileCreated` is available to unsynced gadgets but not to LuaUI afaik. So it's not quite synced but properly unsynced either. See also #1997.

Reply:
Yeah, but also thought maybe that one is wrong, anyways your say on this, I was just pointing out this isn't available for widgets, but as @sprunk said it's also not synced.

Tbh right now it's a bit difficult to know where each callin reaches, like synced/unsynced gadgets, widgets, how does the dev reading the documentation knows?

Reply:
> how does the dev reading the documentation knows?

They don't, I'll try and do a pass on docgen and adjustments to scopes (trying to preserve static analysis types) when I'm able to to fix this issue.

Reply:
> Yeah, but also thought maybe that one is wrong, anyways your say on this, I was just pointing out this isn't available for widgets, but as @sprunk said it's also not synced only.

Ah yeah sorry, in the context of this PR the correction is right. I'm certainly open to having made mistakes when documenting `CLuaHandle`.

> how does the dev reading the documentation knows?

At present I have tried to close this loop here:

https://github.com/beyond-all-reason/RecoilEngine/blob/0683321af751bf99a7fe3fc28f5e1ca76b80c591/rts/Lua/LuaHandle.cpp#L74-L105

Long term I am considering improving docgen by hiding these opaque classnames and instead constructing the various callin scopes via a doc mechanism called "mixins". Which might look something like this:

Input:
```c
/***
 * @class SyncedCallins
 * @include LuaHandle
 * @include LuaHandleSynced
 */
 
/***
 * @mixin LuaHandle
 */

/***
 * @function LuaHandle:ProjectileCreated
 */
```

Output:
```lua
---@class SyncedCallins
SyncedCallins = {}

function SyncedCallins:ProjectileCreated() end
```

Then we can inject these callins into each scope that uses them without ppl having to follow breadcrumbs around the docs.

Still up in the air.

In any case, for now we want to make sure `@see` points to something that is currently defined. It seems LLS check is not verifying these.

Reply:
> This is a copy+paste and could be improved, but that's the place where we'd clarify what the class is and how it's used. It should at least say "contains callins that are available in both synced and unsynced contexts" or similar.

Hmm, but the actual problem is a callin could be available for widgets and/or unsynced gadgets and/or synced gadgets, I know it's not in scope of this PR or conversation to fix that, but just to put the problem forward and try to come up with some solution.

Maybe it should be smth like:

- RulesSyncedCallins: Synced gadget only
- RulesUnsyncedCallins: Unsynced gadget only
- UnsyncedCallins: Unsynced gadget and widget
- Callins: Received by all environments


> In any case, for now we want to make sure `@see` points to something that is currently defined. It seems LLS check is not verifying these.

Yeah I agree, will apply the fix once we settle down the discussion.

Reply:
Yeah, some further thought around how to deal with synced/unsynced scope is required. There was some discussion both in Recoil Discord as well as in the Lua doc thread in BAR discord.

However the general structure you're proposing here is what has been implemented. Actually the truth is that `RulesSyncedCallins` is the callins that are sent to `LuaRules/main.lua`:

https://github.com/beyond-all-reason/RecoilEngine/blob/0683321af751bf99a7fe3fc28f5e1ca76b80c591/rts/Lua/library/RulesSyncedCallins.lua#L3-L5

And `RulesUnsyncedCallins` are the callins sent to `LuaRules/draw.lua`:

https://github.com/beyond-all-reason/RecoilEngine/blob/0683321af751bf99a7fe3fc28f5e1ca76b80c591/rts/Lua/library/RulesUnsyncedCallins.lua#L3-L5

The other classes are just internal types used for composition of the callins and map to internal classes `CLuaHandle`, `CLuaHandleSynced` and `CLuaHandleUnsynced`. My mixin proposal would eliminate these utility classes from the docs (and will be useful for other ways functions/callins are injected into these various scopes).

Reply:
> The whole thing is complicated by Recoil itself has no concept of widget/gadget. In theory these constructs are not even needed to by a game, they can just directly hook into the callins in `draw.lua` and `main.lua`.

Well, recoil does have the concept, but it calls it LuaUI handle (luaui.lua or LuaUI/main.lua), and the split LuaRules handle (LuaRules/main.lua and LuaRules/draw.lua). Ofc gadgets/widgets is just a suborganization of that but in they end they all belong there.

Tbh honest I'm not even sure what combinations are possible, ie, are there callins going just to Synced+UnsyncedLuaRules, are there any that go just to UnsyncedLuaRules?, or do all of those go to UnsyncedLuaUI and UnsyncedLuaRules?